File: examples/rust/transport_http.rs
//! Configure a Wyrd queue with the HTTP fallback transport.
//!
//! Run with:
//!     cargo run -p wyrd-rust-examples --bin transport_http

use wyrd_client::transport::{HttpConfig, QueueConfig, TransportConfig};
use wyrd_spec::security::{SecretRef, TlsConfig};

fn main() -> anyhow::Result<()> {
    let http = HttpConfig {
        base_url: "https://wyrd-ingest.example.com".to_string(),
        timeout_ms: 30_000,
        tls: Some(TlsConfig {
            ca_cert: Some(SecretRef::File {
                path: "/etc/ssl/ca.pem".to_string(),
            }),
            client_cert: None,
            client_key: None,
            server_name_override: None,
            insecure_skip_verify: false,
        }),
        auth: Some(SecretRef::Env {
            name: "WYRD_API_KEY".to_string(),
        }),
        compression: true,
    };
    http.validate()?;

    let queue = QueueConfig {
        transport: TransportConfig::Http(http),
        ..QueueConfig::default()
    };
    queue.validate()?;

    let json = serde_json::to_string_pretty(&queue)?;
    let round_trip: QueueConfig = serde_json::from_str(&json)?;
    anyhow::ensure!(round_trip == queue, "queue config did not round-trip");
    println!("{json}");
    Ok(())
}

File: sdks/wyrd-sdk-python/examples/transport_http.py
"""Configure the HTTP fallback transport as JSON-compatible config.

Run with:
    python sdks/wyrd-sdk-python/examples/transport_http.py
"""

from __future__ import annotations

import json
from typing import Any


def secret_env(name: str) -> dict[str, str]:
    return {"source": "env", "name": name}


def secret_file(path: str) -> dict[str, str]:
    return {"source": "file", "path": path}


def without_none(value: Any) -> Any:
    if isinstance(value, dict):
        return {key: without_none(item) for key, item in value.items() if item is not None}
    return value


def main() -> None:
    http = without_none(
        {
            "base_url": "https://wyrd-ingest.example.com",
            "timeout_ms": 30_000,
            "tls": {
                "ca_cert": secret_file("/etc/ssl/ca.pem"),
                "client_cert": None,
                "client_key": None,
                "server_name_override": None,
                "insecure_skip_verify": False,
            },
            "auth": secret_env("WYRD_API_KEY"),
            "compression": True,
        }
    )
    queue = {
        "transport": {"transport": "http", "params": http},
        "flush_max_rows": 10_000,
        "flush_interval_ms": 5_000,
        "channel_capacity": 100,
    }
    encoded = json.dumps(queue, indent=2, sort_keys=True)
    assert json.loads(encoded) == queue
    print(encoded)


if __name__ == "__main__":
    main()
